Dockerfile

```
@@ -1,112 +1,153 @@
+    --mount=type=cache,target=/root/.cargo/registry,id=fuzz-cargo-registry \
+    --mount=type=cache,target=/root/.rustup,id=fuzz-rustup \
+    --mount=type=cache,target=/root/.gradle,id=fuzz-gradle \
+    --mount=type=cache,target=/root/go/pkg/mod,id=fuzz-go-mod \
+    --mount=type=cache,target=/root/.cache/go-build,id=fuzz-go-build \
+    --mount=type=cache,target=/root/.nuget/packages,id=fuzz-nuget-build \
+    --mount=type=cache,target=/root/.cache/pip,id=fuzz-pip-build \
+    export JAVA_HOME=/usr/lib/jvm/java-21-openjdk-$(dpkg --print-architecture) && \
+    /build/auto_build.sh
+
+FROM base AS runner
+
+LABEL org.opencontainers.image.title="Bitcoin Fuzz"
+LABEL org.opencontainers.image.ref.name="bitcoinfuzz"
+LABEL org.opencontainers.image.description="Bitcoin fuzzing framework for security testing Bitcoin implementations"
+LABEL org.opencontainers.image.source="https://github.com/bitcoinfuzz/bitcoinfuzz"
+LABEL org.opencontainers.image.licenses="MIT"
+LABEL org.opencontainers.image.version="0.1.0"
+LABEL org.opencontainers.image.revision="v2"
 
-# Creates a virtual environment
-RUN python3 -m venv /venv
-ENV PATH="/venv/bin:$PATH"
+WORKDIR /app
 
-# Installs dependencies for embit
-RUN pip install -r modules/embit/requirements.txt
+VOLUME ["/app/data"]
+ARG FUZZ
+ENV FUZZ_DATADIR=/app/data/${FUZZ} \
+    FUZZ=${FUZZ}
 
-# Installs dependencies for C-lightning
-RUN python3 -m pip install --upgrade pip && \
-    python3 -m pip install mako
-RUN git submodule update --init --recursive external/lightning
-RUN apt-get update && apt-get install -y \
-    libsqlite3-dev \
-    libsodium-dev \
-    jq \
-    && rm -rf /var/lib/apt/lists/*
-
-# Clean cache bitcoin core
-RUN rm -rf /app/modules/bitcoin/univalue/build && \
-    mkdir -p /app/modules/bitcoin/univalue/build && \
-    cd /app/modules/bitcoin/univalue && \
-    CXXFLAGS="-fsanitize=address" cmake -B build && \
-    cmake --build build && \
-    cd build && \
-    make univalue
-
-# Copies and gives permission to auto_build.sh
-COPY auto_build.sh .
-RUN chmod +x auto_build.sh
-
-# CMD requires FUZZ, but FUZZ_RUNS and FUZZ_INPUT are optional
-CMD ["bash", "-c", "mkdir -p /app/data/crash && \
-    ./auto_build.sh && \
-    if [ -z \"$FUZZ\" ]; then \
-    echo \"Error: FUZZ not defined\"; \
+SHELL [ "/bin/bash", "-c" ]
+
+RUN if [ -z "${FUZZ}" ]; then \
+    echo "FUZZ build arg var is required"; \
     exit 1; \
-    elif [ -n \"$FUZZ_INPUT\" ]; then \
-    ./bitcoinfuzz -artifact_prefix=/app/data/crash/ \"$FUZZ_INPUT\"; \
-    elif [ -n \"$FUZZ_RUNS\" ]; then \
-    ./bitcoinfuzz -runs=$FUZZ_RUNS -artifact_prefix=/app/data/crash/ /app/data; \
-    else \
-    ./bitcoinfuzz -artifact_prefix=/app/data/crash/ /app/data; \
-    fi"]
+    fi && \
+    useradd -ms /bin/bash bitcoinfuzz && \
+    mkdir -p ${FUZZ_DATADIR} && \
+    chown -R bitcoinfuzz:bitcoinfuzz /app/data
+USER bitcoinfuzz
+
+COPY --from=builder --chmod=0755 /build/bitcoinfuzz .
+COPY --from=builder /build/bitcoinfuzz/*.so .
```

The shared libraries will be inside the build folder

```suggestion
COPY --from=builder /build/*.so .
```